src/absorb.rs

```diff
@@ -0,0 +1,364 @@
+pub trait AbsorbWithLength: Absorb {
+    /// The length of the `self` being absorbed.
+    fn absorb_length(&self) -> usize;
+
+    /// Converts the object into a list of bytes along with its length information
+    /// that can be absorbed by a `CryptographicSponge`.
+    /// Append the list to `dest`.
+    fn to_sponge_bytes_with_length(&self, dest: &mut Vec<u8>) {
+        self.absorb_length().to_sponge_bytes(dest);
+        self.to_sponge_bytes(dest)
+    }
+
+    /// Converts the object into field elements along with its length information
+    /// that can be absorbed by a `CryptographicSponge`.
+    /// Append the list to `dest`
+    fn to_sponge_field_elements_with_length<F: PrimeField>(&self, dest: &mut Vec<F>) {
+        self.absorb_length().to_sponge_field_elements(dest);
+        <Self as Absorb>::to_sponge_field_elements(&self, dest)
+    }
+}
+
+/// If `F1` equals to `F2`, return `x` as `F2`, otherwise panics.
+/// ## Panics
+/// This function will panic if `F1` is not equal to `F2`.
+pub(crate) fn field_cast<F1: PrimeField, F2: PrimeField>(input: F1) -> F2 {
+    if TypeId::of::<F1>() != TypeId::of::<F2>() {
+        panic!("Trying to absorb non-native field elements.")
+    } else {
+        F2::from_le_bytes_mod_order(&to_bytes!(input).unwrap())
+    }
+}
+
+/// If `F1` equals to `F2`, add all elements of x as `F2` to `dest`, otherwise panics.
+/// ## Panics
+/// This function will panic if `F1` is not equal to `F2`.
+pub(crate) fn batch_field_cast<F1: PrimeField, F2: PrimeField>(x: &[F1], dest: &mut Vec<F2>) {
+    if TypeId::of::<F1>() != TypeId::of::<F2>() {
+        panic!("Trying to absorb non-native field elements.")
+    } else {
+        x.iter()
+            .for_each(|item| dest.push(F2::from_le_bytes_mod_order(&to_bytes!(item).unwrap())))
+    }
+}
+
+impl Absorb for u8 {
+    fn to_sponge_bytes(&self, dest: &mut Vec<u8>) {
+        self.write(dest).unwrap()
+    }
+
+    fn to_sponge_field_elements<F: PrimeField>(&self, dest: &mut Vec<F>) {
+        dest.push(F::from(*self))
+    }
+
+    fn batch_to_sponge_bytes(batch: &[Self], dest: &mut Vec<u8>) {
+        dest.extend_from_slice(batch)
+    }
+
+    fn batch_to_sponge_field_elements<F: PrimeField>(batch: &[Self], dest: &mut Vec<F>) {
+        let mut bytes = (batch.len() as u64).to_le_bytes().to_vec();
+        bytes.extend_from_slice(batch);
+        dest.extend_from_slice(&bytes.to_field_elements().unwrap()[..])
+    }
+}
+
+impl Absorb for bool {
+    fn to_sponge_bytes(&self, dest: &mut Vec<u8>) {
+        dest.push(*self as u8)
+    }
+
+    fn to_sponge_field_elements<F: PrimeField>(&self, dest: &mut Vec<F>) {
+        dest.push(F::from(*self))
+    }
+}
+
+macro_rules! impl_absorbable_field {
+    ($field:ident, $params:ident) => {
+        impl<P: $params> Absorb for $field<P> {
+            fn to_sponge_bytes(&self, dest: &mut Vec<u8>) {
+                self.write(dest).unwrap()
```

Comment:
Let's use `CanonicalSerialize` for this, and let's avoid `to_bytes`/`ToBytes` in general, wherever possible.

Reply:
fixed (I replaced all occurrences of `to_bytes` to serialize)